R/utils-obs-unit.R

```diff
@@ -16,35 +16,134 @@ task_id_cols_to_validate <- function(tbl, config_tasks) {
+  ) {
+    obs_unit <- c(obs_unit, "as_of")
+  }
+
+  obs_unit
+}
+
+#' Get observation unit columns by inferring from table and task configuration.
+#'
+#' Infers the observation unit from the table by combining task identifier
+#' columns from the hub config with `"as_of"` (if present in the table).
+#' This approach is used for backward compatibility when target-data.json
+#' configuration is not available.
+#'
+#' @param tbl A target data frame (oracle output or time-series target data).
 #' @param config_tasks A list of task configurations from the hub config.
 #' @return A character vector of columns that define the observation unit.
 #' @noRd
-get_tbl_obs_unit <- function(tbl, config_tasks) {
+get_obs_unit_from_tbl <- function(tbl, config_tasks) {
+  # Inference behavior: use task IDs that are present in the table
   obs_unit <- task_id_cols_to_validate(tbl, config_tasks)
+
+  # Always include as_of if present in the table (backward compatibility)
   if ("as_of" %in% colnames(tbl)) {
     obs_unit <- c(obs_unit, "as_of")
   }
+
   obs_unit
 }
 
-#' Group a data table by observation unit columns.
+#' Get target data observation unit columns.
+#'
+#' Determines the observation unit for target data, which defines the unique
+#' combinations of columns that identify each independent observation.
 #'
-#' Uses `get_tbl_obs_unit()` to determine grouping columns based on task
-#' configuration and the presence of `"as_of"`. Returns the input table grouped
-#' by those columns, suitable for downstream validation or summarisation.
+#' When `config_target_data` is provided, uses `get_obs_unit_from_config()`
+#' to extract the observable unit from the target-data.json configuration.
 #'
-#' @param tbl A data frame (e.g., oracle output target data or model output).
+#' When `config_target_data` is `NULL`, uses `get_obs_unit_from_tbl()` to
+#' infer the observation unit from task identifier columns and the presence
+#' of `as_of` in the table.
+#'
+#' @param tbl A target data frame (oracle output or time-series target data).
 #' @param config_tasks A list of task configurations from the hub config.
+#' @param config_target_data Optional target-data.json config object. When `NULL`,
+#'   falls back to inference behavior based on task IDs.
+#' @param target_type Optional target type: `"time-series"` or `"oracle-output"`.
+#'   Required when `config_target_data` is provided.
+#' @param include_as_of Logical indicating whether to include `as_of` in the
+#'   observation unit. When `config_target_data` is provided, `as_of` is only
+#'   added if this is `TRUE` AND the data is versioned (per
+#'   `hubUtils::get_versioned()`). When `config_target_data` is `NULL`, the
+#'   inference behavior is used (always include `as_of` if present in the table),
+#'   regardless of this parameter's value.
+#' @return A character vector of columns that define the observation unit.
+#' @noRd
+get_obs_unit <- function(
+  tbl,
+  config_tasks,
```

Comment:
Minor suggestion/question: as this parameter is only use in `get_obs_unit_from_tbl` can we set to `NULL` if `config_target_data` is NOT set to `NULL`? As it's an internal function I don't think we really need to worry about it, I was just wondering if it's worth adding that information somewhere?